src/index.ts

```diff
@@ -0,0 +1,198 @@
+	return async function fetchWrap(url: string, opts: FetchOptions = {}): Promise<Response> {
+		// @ts-ignore
+		if (!opts.agent) {
+			// Add default `agent` if none was provided
+			// @ts-ignore
+			opts.agent = getAgent(url, { AGENT_OPTIONS, ...agentOpts });
+		}
+
+		opts.redirect = 'manual';
+
+		opts.headers = new Headers(opts.headers);
+		if (!(opts.headers instanceof Headers)) {
+			throw new Error('Failed to create fetch opts');
+		}
+
+		// Workaround for node-fetch + agentkeepalive bug/issue
+		const parsedUrl = parseUrl(url);
+		const host = opts.headers.get('host') || parsedUrl.host;
+
+		if (!host) {
+			throw new TypeError('Unable to determine Host');
+		}
+
+		opts.headers.set('host', host);
+
+		const ip = isIP(parsedUrl.hostname || '');
+		if (ip === 0) {
+			if (!parsedUrl.hostname) {
+				throw new Error('Unable to determine hostname');
+			}
+
+			parsedUrl.hostname = await resolve(parsedUrl.hostname);
+			url = formatUrl(parsedUrl);
+		}
+
+		// Convert Object bodies to JSON
+		if (opts.body && typeof opts.body === 'object' && !(Buffer.isBuffer(opts.body) || opts.body instanceof Readable)) {
+			opts.body = JSON.stringify(opts.body);
+			opts.headers.set('Content-Type', 'application/json');
+			opts.headers.set('Content-Length', `${Buffer.byteLength(opts.body)}`);
+		}
+
+		const retryOpts = Object.assign({
+			// timeouts will be [ 10, 50, 250 ]
+			minTimeout: MIN_TIMEOUT,
+			retries: MAX_RETRIES,
+			factor: FACTOR,
+		}, opts.retry);
+
+		if (opts.onRetry) {
+			// Using `any` here to avoid type mismatch between
+			// Error and FetchRetryError
+			retryOpts.onRetry = (error: any) => {
+				if (opts.onRetry) {
+					opts.onRetry(error, opts);
+					if (opts.retry && opts.retry.onRetry) {
+						opts.retry.onRetry(error);
+					}
+				}
+			}
+		}
+
+		debug('%s %s', opts.method || 'GET', url);
+		const res = await retry(async (_bail, attempt) => {
+			const isRetry = attempt < retryOpts.retries;
+
+			try {
+				const res = await fetch(url, opts);
+
+				debug('status %d', res.status);
+				if (res.status >= 500 && res.status < 600 && isRetry) {
+					throw new FetchRetryError(url, res.status, res.statusText);
+				} else {
+					return res;
+				}
+			} catch (err) {
+				const { method = 'GET' } = opts;
+				debug(`${method} ${url} error (${err.status}). ${isRetry ? 'retrying' : ''}`, err);
+				throw err;
+			}
+		}, retryOpts);
+
+		if (isRedirect(res.status)) {
+			const redirectOpts = Object.assign({}, opts);
+			redirectOpts.headers = new Headers(opts.headers);
+
+			// per fetch spec, for POST request with 301/302 response, or any request with 303 response, use GET when following redirect
+			if (
+				res.status === 303 ||
+				((res.status === 301 || res.status === 302) && opts.method === 'POST')
+			) {
+				redirectOpts.method = 'GET';
+				redirectOpts.body = undefined;
+				redirectOpts.headers.delete('content-length');
+			}
+
+			const location = res.headers.get('Location');
```

Comment:
```suggestion
			const location = res.headers.get('location');
```

Should we lower-case all headers? now it's inconsistent.